salt/_runners/metalk8s_checks.py

```diff
@@ -0,0 +1,246 @@
+
+    not_ready_nodes = {}
+    for node_name in node_list:
+        node_obj = __salt__["salt.cmd"](
+            fun="metalk8s_kubernetes.get_object",
+            kind="Node",
+            apiVersion="v1",
+            name=node_name,
+        )
+
+        condition = next(
+            cond for cond in node_obj["status"]["conditions"] if cond["type"] == "Ready"
+        )
+        if condition["status"] != "True":
+            not_ready_nodes.setdefault(condition["reason"], []).append(node_name)
+
+    for reason, nodes_names in not_ready_nodes.items():
+        errors.append(
+            "Nodes '{}' are not ready - {}".format("', '".join(nodes_names), reason)
+        )
+
+    return _handle_errors(errors, raises)
+
+
+def minions(minion_list=None, raises=True):
+    """Check that all Salt minions are Ready
+
+    Args:
+        minion_list (list, optional): List of Salt minion to check. Defaults to `pillar.metalk8s.nodes`.
+        raises (bool, optional): Whether or not this function should raise. Defaults to True.
+
+    Raises:
+        CheckError: If 'raises' is True and some Salt minions are not ready
+
+    Returns:
+        dict or True: An error message or True if everything OK
+    """
+    errors = []
+
+    if minion_list is None:
+        minion_list = __pillar__["metalk8s"]["nodes"].keys()
+
+    ret = __salt__["salt.execute"](
+        ",".join(minion_list), "test.ping", tgt_type="list", timeout=10
+    )
+
+    not_ready_minions = {
+        "are not ready": [
+            minion for minion, minion_ret in ret.items() if minion_ret is not True
+        ],
+        "did not answered": list(set(minion_list) - set(ret.keys())),
+    }
+
+    for reason, minions_names in not_ready_minions.items():
+        if minions_names:
+            errors.append(
+                "Salt minions '{}' {}".format("', '".join(minions_names), reason)
+            )
+
+    return _handle_errors(errors, raises)
+
+
+def upgrade(dest_version, saltenv, raises=True):
+    """Check that we can start MetalK8s cluster upgrade
+
+    Args:
+        dest_version (string): Destination version for upgrade
+        saltenv (string): Salt environment that will be used for upgrade
+        raises (bool, optional): Whether or not this function should raise. Defaults to True.
+
+    Raises:
+        CheckError: If 'raises' is True and some upgrade condition are not met
+
+    Returns:
+        dict or True: An error message or True if everything OK
+    """
+    errors = []
+
+    # NOTE: We use `get_from_map` as we want also defaults from `map.jinja`
+    # to be merged
+    metalk8s_pillar = __salt__["salt.cmd"](
+        "metalk8s.get_from_map", "metalk8s", saltenv=saltenv, with_pillar=True
+    )
+
+    # When upgrading the saltenv should be the "metalk8s-<destination version>"
+    if saltenv != f"metalk8s-{dest_version}":
+        errors.append(
+            f"Invalid saltenv '{saltenv}' consider using 'metalk8s-{dest_version}'"
+        )
```

Comment:
Why bothering to pass the `saltenv` if it should match the `dest_version` anyway, shouldn't we just use one or the other ?

Reply:
It's mainly because this one is used in the upgrade orchestrate where you must have saltenv, and dest_version come from the `kube-system` annotation so we need to check it